sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -474,6 +474,20 @@ case class MapObjects private(
           s"$seq == null ? $array[$loopIndex] : $seq.apply($loopIndex)"
     }
 
+    // Make a copy of the data if it's unsafe-backed
+    val genFunctionValue = lambdaFunction.dataType match {
+      case StructType(_) =>
+        s"(${genFunction.value} instanceof ${classOf[MutableRow].getName}? " +
```

Comment:
done.
indeed we should narrow down to `UnsafeRow`, since other `MutableRow`s' backed data are not shared even though they are mutable.